_posts/2017-08-31-Rust-1.20.md

```diff
@@ -0,0 +1,261 @@
+}
+```
+
+These are called "associated functions" becuase they are functions that are
+associated with the type, that is, they're attached to the type itself, and
+not any particular instance.
+
+Rust 1.20 adds the ability to define "associated constants" as well:
+
+```rust
+struct Struct;
+
+impl Struct {
+    const ID: u32 = 0;
+}
+
+fn main() {
+    println!("the ID of Struct is: {}", Struct::ID);
+}
+```
+
+That is, the constant `ID` is associated with `Struct`. Like functions,
+associated constants work with traits and enums as well.
+
+Traits have an extra ability with associated constants that gives them some
+extra power. With a trait, you can use an associated constant in the same way
+you'd use an associated type: by declaring it, but not giving it a value. The
+implementor of the trait then declares its value upon implementation:
+
+```rust
+trait Trait {
+    const ID: u32;
+}
+
+struct Struct;
+
+impl Trait for Struct {
+    const ID: u32 = 5;
+}
+
+fn main() {
+    println!("{}", Struct::ID);
+}
+```
+
+Before this release, if you wanted to make a trait that represented floating
+point numbers, you'd have to write this:
+
+```rust
+trait Float {
+    fn nan() -> Self;
+    fn infinity() -> Self;
+    ...
+}
+```
+
+This is slightly unweildy, but more importantly, becuase they're functions, they
+cannot be used in constant expressions, even though they only return a constant.
+Because of this, a design for `Float` would also have to include constants as well:
+
+```rust
+mod f32 {
+    const NAN: f32 = 0.0f32 / 0.0f32;
+    const INFINITY: f32 = 1.0f32 / 0.0f32;
+
+    impl Float for f32 {
+        fn nan() -> Self {
+            f32::NAN
+        }
+        fn infinity() -> Self {
+            f32::INFINITY
+        }
+    }
+}
+```
+
+Associated constants let you do this in a much cleaner way. This trait definition:
+
+```rust
+trait Float {
+    const NAN: Self;
+    const INFINITY: Self;
+    ...
+}
+```
+
+Leads to this implementation:
+
+```rust
+mod f32 {
+    impl Float for f32 {
+        const NAN: f32 = 0.0f32 / 0.0f32;
+        const INFINITY: f32 = 1.0f32 / 0.0f32;
+    }
+}
+```
+
+much cleaner, and more versitile.
+
+Associated constants were proposed in [RFC 195], almost exactly three years ago. It's
+been quite a while for this feature! That RFC contained all associated items, not just
+constants, and so some of them, such as associated types, were implemented faster than
+others. In general, we've been doing a lot of internal work for constant evaluation,
+to increase Rust's capabilities for compile-time metaprogramming. Expect more on this
+front in the future.
+
+[RFC 195]: https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md
+
+Back in [Rust 1.14], we announced preliminary support for asm.js and wasm
+with the help of Emscripten. Since then, LLVM has added its own support, and
+so, we have [added native wasm support to Rust]! It's using the
+`wasm32-experimental-emscripten` target.
```

Comment:
I think that's just what we decided to call it, not what the LLVM target is actually called.

Reply:
Yes, that was my intention. Any idea of how to word it better?

Reply:
This new target is named `wasm32-experimental-emscripten`.

Reply:
When mentioning this I think we'd just want to set expectations by saying that it's an experimental target and doesn't even have nightly support right now. Furthermore this isn't even using *our* LLVM's backend support for WebAssembly but rather we still ship IR to emscripten (AFAIK) and then emscripten uses LLVM's WebAssembly backend.

All in all this may not be worth calling out :(